Use PotentialNamespace in ancestry

Allda opened the pull request:
If PotentialNamespace is available for a given feature it is used instead
of detector namespace. A detector is empty in such a case.

Resolves: https://github.com/coreos/clair/issues/734

Allda committed:
Use PotentialNamespace in ancestry

If PotentialNamespace is available for given feature it is used instead
of detector namespace. Detector is empty in such a case.

Allda commented:
@KeyboardNerd I tested the PR with real data which uses PotentialNamespace and results were exactly as I expected.

KeyboardNerd commented on ancestry.go:
```diff
@@ -211,9 +211,20 @@ func (b *AncestryBuilder) createLayerIndexedFeature(namespace *layerIndexedNames
+			},
+			IntroducedIn: b.layerIndex,
+		}
+		matchedNamespaces = append(matchedNamespaces, a)
```
We should always ensure that all elements in matched namespaces are in the ancestrybuilder.namespaces, and any feature with the same namespace must point to the same namespace object in the ancestrybuilder.namespaces slice. 

Otherwise, this will break the updateNamespace logic. 

BTW, in order to use the CPE string as a namespace, please remove the panic condition for updateNamespace logic.

KeyboardNerd replied:
^^^ 
This may not be the case.
I think for now at least: when a feature has a potential namespace, we will use that namespace directly instead of getting into the namespace lookup and upgrade process. So that the output AncestryFeature is guaranteed to have the namespace.

@jzelinskie What do you think about this logic here?

jzelinskie replied:
I think this logic is fine for now. I don't want to put the cart before the horse. Let's get things working and then test our assumptions later.

KeyboardNerd reviewed (changes requested):
Please also update the convert.go DetectorFromDatabaseModel function. Right now I think the API may fail because the detector is empty.

Allda commented:
> Please also update the convert.go DetectorFromDatabaseModel function. Right now I think the API may fail because the detector is empty.

API doesn't fail when a detector is empty.

Here is an example output:
`
{
  "name": "gnupg2",
  "namespace": {
    "name": "cpe:/a:redhat:cloudforms_managementengine:5.8::el7",
    "detector": {}
  },
  "version": "2.0.22-5.el7_5",
  "version_format": "rpm",
  "detector": {
    "name": "redhat",
    "version": "1.0",
    "dtype": "DETECTOR_D_TYPE_FEATURE"
  },
  "feature_type": "binary"
},`

Allda committed:
Convert: return nil when detector is empty

When PotentialNamespace is used detector is empty and
DetectorFromDatabaseModel() should return nil instead of empty object.
